### ISSUE TITLE
Troublesome Build

### ISSUE THREAD

**Issue by mabbamOG** [koreader/koreader#2798]
* KOReader version: latest git as of a couple of hours ago
* Device: archlinux on x86_64

#### Issue
the building process is _very_ painful. and it does not result in a working binary.
so the initial clone is endless download.
the fetch-third-party is endless downloads.
the build is also endless downloads (like why?) of libs i already have running on my pc and clones everywhere. not clean at all
the final run has some more compilation (wtf) and then it crashes...the attached file is the final part of the log

#### Steps to reproduce
1. check that the programs mentioned in the readme are installed (no install commands, just which them out)
2. git clone, bootstrap and then build. honestly this process is dirty as hell there's loads of stuff being downloaded
3. kodev run

**Comment by mabbamOG** [comment]
(the crash is immediate. you can see a window for a split second..)

**Comment by houqp** [comment]
I totally understand why you are getting frustrated ;) It's really easy for people to think that koreader is just a simple reader, how complex can it be.

Well, it's in fact closer to a bundled Linux distribution for a reader that runs on its own UI framework that interacts directly with various kernel interfaces. See https://github.com/koreader/koreader-base/tree/master/thirdparty for extra fun.

The reason for having all these user space libraries bundled is we have users running koreader on all kinds of different platforms including kindle, kobo, pocketbook, android and other random models. All of these models ship different userspaces. Even amazon ships different userspaces on their own devices. So how do you create a single code base that works on all platform consistently? You bundle as many dependencies as you can. I think we have bundled almost everything except libc at this point. And yes, those different libc versions out there are still causing us headaches and we had to put into hacks to work around those issues.

Another reason is most of the devices are shipped with really old userspaces, which is not even compatible with the rendering engine that we are using like mupdf and crengine. So we have to include those dependencies and link to them at build time using `rpath`.

We can surely just use all the existing user space libraries shipped within each distribution for development purpose. But then whatever you developed on won't match with what's actually running the user device. And that's how you get bugs that can't be reproduced on emulator due to differences in userspace. It's a tradeoff we made to sacrifice a bit of developer happiness for a better user experience. It's easier to produce a more predictable program when your development environment is closer to production environment.

I know not everyone likes this dependency building game. In fact I hated it myself. That's why I prepared both vagrant and docker images to help other people recreate the exact development environment with one command. See: https://github.com/koreader/virdevenv.

Lastly, this emulator build process is not officially tested under Arch. One thing I do know is cross compiling definitely won't work under Arch due to nuances on Arch's build configs.

**Comment by houqp** [comment]
At the mean time, if you are interested in getting the emulator working on arch, you are well come to give it a try. I am happy to help where ever I can.

If you already have the build process worked out but emulator is crashing on start, then please paste the log from your terminal. It should include a stacktrace.

**Comment by Frenzie** [comment]
> I know not everyone likes this dependency building game. In fact I hated it myself.

Oh, it's awful, but I wouldn't say you're sacrificing developer happiness by making it as close to my actual device as possible. What sacrifices my happiness is having to use USBNet. :-P

**Comment by mabbamOG** [comment]
Thanks for the informative reply! that certainly explains the quite odd behavious of koreader.
I will be happy to test koreader out on archlinux once i get hacking on a kindle myself.

How would you fare the difficulty of adding a new "online source" to the reader? something like maybe pocket with a synced list and everything...i can envision creating a standalone app (pygtk) for that but im not sure how i would appropriately render an article on an ereader (appropriately meaning it looks good like normal kindle articles or kobo)

**Comment by Frenzie** [comment]
Also see #1580 and #2621. I shouldn't think it's any harder than Python. For rendering you'd simply use crengine or MuPDF, the simplest way obviously being just to use files. Of course you could make things a lot more complicated with an SQLite DB, a nice GUI, etc. but as a start I think something in the general vain of #2592 would make people plenty happy.

**Comment by mabbamOG** [comment]
adding `./kodev run` log. also have the build log if you need it!
[log.txt](https://github.com/koreader/koreader/files/956803/log.txt)

**Comment by Frenzie** [comment]
That would probably be helpful. The JSON module comes from luajson and I think it should normally live (for the emulator) where you get `no file 'rocks/lib/lua/5.1/json.so'`.

**Comment by mabbamOG** [comment]
`pwd`:
```
/home/username/Documents/000/koreader/base/build/x86_64-pc-linux-gnu/rocks/share/lua/5.3
```

`tree`:
```
.
├── json
│   ├── decode
│   │   ├── composite.lua
│   │   ├── number.lua
│   │   ├── others.lua
│   │   ├── state.lua
│   │   ├── strings.lua
│   │   └── util.lua
│   ├── decode.lua
│   ├── encode
│   │   ├── array.lua
│   │   ├── calls.lua
│   │   ├── number.lua
│   │   ├── object.lua
│   │   ├── others.lua
│   │   ├── output.lua
│   │   ├── output_utility.lua
│   │   └── strings.lua
│   ├── encode.lua
│   └── util.lua
├── json.lua
├── re.lua
├── Spore
│   ├── Core.lua
│   ├── GoogleDiscovery.lua
│   ├── Middleware
│   │   ├── Auth
│   │   │   ├── AWS.lua
│   │   │   ├── Basic.lua
│   │   │   ├── Bearer.lua
│   │   │   ├── DataPublica.lua
│   │   │   ├── Digest.lua
│   │   │   └── OAuth.lua
│   │   ├── Cache.lua
│   │   ├── DoNotTrack.lua
│   │   ├── Format
│   │   │   ├── JSON.lua
│   │   │   ├── XML.lua
│   │   │   └── YAML.lua
│   │   ├── Logging.lua
│   │   ├── Mock.lua
│   │   ├── Parameter
│   │   │   ├── Default.lua
│   │   │   └── Force.lua
│   │   ├── Proxy
│   │   │   └── Basic.lua
│   │   ├── Redirection.lua
│   │   ├── Runtime.lua
│   │   └── UserAgent.lua
│   ├── Protocols.lua
│   ├── Request.lua
│   ├── WADL.lua
│   └── XML.lua
└── Spore.lua

9 directories, 45 files
```

and build log is attached.

[buildlog.txt](https://github.com/koreader/koreader/files/959278/buildlog.txt)

**Comment by Frenzie** [comment]
In that case you should be able to run it by changing 5.1 to 5.3 here: https://github.com/koreader/koreader/blob/00fbc5f3c9574e3f63faac4dee45b0c7dcf7d8d3/setupkoenv.lua#L3-L6 But it's possible that you'd have to build luarocks with `--lua-suffix=5.1` or some such because LuaJIT is compatible with the Lua 5.1 API+ABI.

This is actually the reason I haven't closed https://github.com/koreader/koreader/issues/2613#issuecomment-285911838 yet:

> @chrox and I also talked about building self-contained luarocks from source for dev environemnt so we can have full control over the lua dependency and make sure everything is setup correctly.

PS Actually [the Travis build already does that](https://github.com/koreader/koreader/blob/96ea2ece96d90094505d24e2934f71eeb6c6d588/.ci/install.sh#L25-L31).

**Comment by z3t0** [comment]
Is it just me or does setting up the vm take a very long time? Vagrant up was fine but running ./koreader run is taking a very long time... It's been about 1.5 hrs. Is this normal?

**Comment by Frenzie** [comment]
Depends on the speed of your connection and computer. It should tell you what it's doing? :)

**Comment by z3t0** [comment]
It seems like its just downloading and compiling a bunch of dependencies ... feels like its building its own operating system haha

**Comment by Frenzie** [comment]
Yup, kind of.

**Comment by mabbamOG** [comment]
> In that case you should be able to run it by changing 5.1 to 5.3 here: https://github.com/koreader/koreader/blob/00fbc5f3c9574e3f63faac4dee45b0c7dcf7d8d3/setupkoenv.lua#L3-L6 

I changed 5.1 to 5.3 and it seems to be working!
The startup is a little slow (a couple of seconds) and there's lots of output, but it seems to work fine.

There is one small bug which is annoying on the kindle: deleting a repository is not possible. Instead of making it unfeasible, in my opinion it would be better to show a confirmation dialogue instead

**Comment by houqp** [comment]
It's highly recommended to get a proper luajit installation for luarocks instead of changing it to 5.3 since 5.3 is not compatible with LUAJIT(5.1).

The startup is slow because `kodev run` by default always run the `make` command to keep the build update date before starting the emulator. You can use `kodev run --no-build` to skip the build process.

> There is one small bug which is annoying on the kindle: deleting a repository is not possible. Instead of making it unfeasible, in my opinion it would be better to show a confirmation dialogue instead

What do you mean by repository? A file folder?

**Comment by mabbamOG** [comment]
yes i mixed up the terms. i meant directory :)

**Comment by houqp** [comment]
Yeah, you can only delete empty directory right now. Feel free to send a PR for recursive directory deletion.

**Comment by iwanb** [comment]
I was also trying to make the emulator run on ArchLinux, this seems to do the trick:
 - Install luarocks5.1 so you have a version compatible with LuaJIT
 - In the file kodev modify eval "$(luarocks path --bin)" by eval "$(luarocks-5.1 path --bin)"
 - Run ./kodev activate

Maybe the luarocks executable could be taken from an environment variable to be flexible?

**Comment by Frenzie** [comment]
@iwanb Good idea, you can make a PR if you like or else I'll take care of it at some indeterminate time in the future. :-)

**Comment by pazos** [comment]
I have a long-term proposal of removing the build steps of the README.sh (including prerrequisites) and having new -focused- build steps for different targets: sdl, eink devices and android.

**Comment by Frenzie** [comment]
The section could certainly be rewritten/reorganized. Since device/Android is basically what you need to do on top of SDL, I'd propose moving those parts of the "prerequisite" information further down into the dedicated device/Android instructions. That would automatically streamline the information about the basic emulator build process.

**Comment by Frenzie** [comment]
I'll close this as mostly resolved by #5217, making this ticket too broad. Some information still needs to be added about how to utilize Docker, but I already wrote a stub over in https://github.com/koreader/virdevenv a little while ago.
